app/code/Magento/Inventory/Model/Source/Validator/CarrierLinks.php

```diff
@@ -20,31 +21,49 @@ class CarrierLinks implements SourceValidatorInterface
     private $validationResultFactory;
 
     /**
+     * Shipping config
+     *
+     * @var Config
+     */
+    private $shippingConfig;
+
+    /**
      * @param ValidationResultFactory $validationResultFactory
      */
-    public function __construct(ValidationResultFactory $validationResultFactory)
+    public function __construct(ValidationResultFactory $validationResultFactory, Config $shippingConfig)
     {
         $this->validationResultFactory = $validationResultFactory;
+        $this->shippingConfig = $shippingConfig;
     }
 
     /**
      * @inheritdoc
      */
     public function validate(SourceInterface $source): ValidationResult
     {
-        $value = $source->getCarrierLinks();
+        $carrierLinks = $source->getCarrierLinks();
 
         $errors = [];
-        if (null !== $value) {
-            if (!is_array($value)) {
+        if (null !== $carrierLinks) {
+            if (!is_array($carrierLinks)) {
                 $errors[] = __('"%field" must be list of SourceCarrierLinkInterface.', [
                     'field' => SourceInterface::CARRIER_LINKS
                 ]);
-            } else if (count($value) && $source->isUseDefaultCarrierConfig()) {
+            } elseif (count($carrierLinks) && $source->isUseDefaultCarrierConfig()) {
                 $errors[] = __('You can\'t configure "%field" because you have chosen Global Shipping configuration.', [
                     'field' => SourceInterface::CARRIER_LINKS
                 ]);
             }
+
+            $availableCarriers = $this->shippingConfig->getAllCarriers();
```

Comment:
Looks like we should refactor this logic
Becuase if `$carrierLinks` is not array then we do not need to perform `foreach`